controllers/file.go

```diff
@@ -0,0 +1,261 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"path"
+
+	"github.com/casvisor/casvisor/object"
+	"github.com/casvisor/casvisor/storage"
+	"github.com/casvisor/casvisor/util"
+)
+
+// UpdateFile
+// @Title UpdateFile
+// @Tag File API
+// @Description update file
+// @Param sessionId query string true "The id of the session"
+// @Param filename query string true "The name of the file"
+// @Success 200 {object} controllers.Response The Response object
+// @router /update-file [post]
+func (c *ApiController) UpdateFile() {
+	c.ResponseOk()
+}
+
+// AddFile
+// @Title AddFile
+// @Tag File API
+// @Description add file
+// @Param key query string true "The directory of the file"
+// @Param sessionId query string true "The id of the session"
+// @Success 200 {object} controllers.Response The Response object
+// @router /add-file [post]
+func (c *ApiController) AddFile() {
+	userName := c.GetSessionUsername()
+	sessionId := c.Input().Get("id")
+	key := c.Input().Get("key")
+	isLeaf := c.Input().Get("isLeaf") == "1"
+
+	session, err := object.GetConnSession(sessionId)
+	if err != nil {
+		c.ResponseError(err.Error())
+		return
+	}
+	if session == nil {
+		c.ResponseError("session not found")
+		return
+	}
+
+	provider, err := storage.GetStorageProvider(session.Protocol, sessionId, "")
+	if err != nil {
+		c.ResponseError(err.Error())
+		return
+	}
+
+	if isLeaf {
+		_, file, err := c.GetFile("file")
+		if err != nil {
+			c.ResponseError(err.Error())
+			return
+		}
+
+		srcFile, err := file.Open()
+		if err != nil {
+			c.ResponseError(err.Error())
+			return
+		}
+
+		var fileBuffer *bytes.Buffer
+		fileBuffer = bytes.NewBuffer(nil)
+		_, err = io.Copy(fileBuffer, srcFile)
+		if err != nil {
+			c.ResponseError(err.Error())
+			return
+		}
+
+		_, err = provider.PutObject(userName, key, file.Filename, fileBuffer)
+		if err != nil {
+			c.ResponseError(err.Error())
+			return
+		}
+
+		addRecordForFile(c, userName, "Upload", sessionId, key, file.Filename, false)
+		c.ResponseOk()
+	} else {
+		err = provider.Mkdir(key)
+		if err != nil {
+			c.ResponseError(err.Error())
+			return
+		}
+
+		addRecordForFile(c, userName, "MkDir", sessionId, key, "", false)
+		c.ResponseOk()
+	}
+}
+
+// DownloadFile
+// @Title DownloadFile
+// @Tag File API
+// @Description download file
+// @Param sessionId query string true "The id of the session"
+// @Param key query string true "The direction of the file"
+// @Success 200 {object} controllers.Response The Response object
+// @router /get-file [get]
+func (c *ApiController) DownloadFile() {
```

Comment:
Use DownloadFile instead of getFile because getFile is the beego framework built in function.

Reply:
Got it